iterate on the new BinaryTool subsystem and use it to create a new LLVM subsystem

cosmicexplorer opened the pull request:
### Problem/Solution

I'm working on providing a "native" (C/C++) compilation toolchain with BinaryUtil so that we can build `python_dist()` targets in a controlled environment, and so users don't have to provide all the elements of the toolchain on every machine they might want to run Pants on.

In #5443, @benjyw provided a very slick way to define subsystems which provide downloaded files using BinaryUtil. This PR iterates on that and uses it to define an example "LLVM" subsystem, which does not yet exist in BinaryUtil, but shows how simple it is to compose elements of the BinaryTool class.

### Result

It is now possible to compose supportdirs to easily make families of subsystems by overriding `support_dir_paths()`, as is done in `NativeTool` and `Script`, and which is consumed by the `Protoc` and `LLVM` subsystems.

It is also now possible to very easily provide a compressed archive using BinaryTool or BinaryUtil, for platform-dependent code or scripts, using the archivers defined in `fs/archive.py`, by specifying e.g. `archive_type = 'tgz'` in a subclass of BinaryTool. This makes it much easier to provide complex toolchains, such as a native code compilation toolchain.

I'm preparing another PR on top of this, which consumes the LLVM subsystem and plumbs it into `BuildLocalPythonDistributions` to compile native code specified in a `python_dist()` target.

benjyw commented on src/python/pants/binaries/binary_tool.py:
```diff
@@ -32,6 +35,17 @@ class BinaryToolBase(Subsystem):
   extra_version_option_kwargs = None
 
+  class InvalidSupportDir(TaskError):
+    """Indicates that the subclass of BinaryToolBase did not set up a valid
```
Nit: The first sentence of a docstring should fit on a single line.

I think you can convey this in half the number of words.

benjyw replied:
Also, let's be consistent with the naming of "support_dir" vs. "support_subdir". BinaryUtil uses "supportdir", but it's ok to be inconsistent with that here, since we're going to make BinaryUtil into an implementation detail and not allow direct use of it. But at least within this file let's be consistent.

cosmicexplorer replied:
I have removed the mess of joining directory paths and made BinaryTool subclasses go under `bin/`, so I have removed this exception and the naming confusion, as they are no longer needed. I will keep your comments in mind for this and future PRs, though!

benjyw commented on src/python/pants/binaries/binary_tool.py:
```diff
@@ -70,10 +84,35 @@ def select(self, context=None):
+  def _binary_util(self):
+    return BinaryUtil.Factory.create()
+
+  # can override this and call super() to compose
```
Nit: Complete sentences imply complete thoughts. So begin a sentence with a capital letter, end it with a period.

cosmicexplorer replied:
This was also removed in 27d0c37, but I will also keep this in mind.

benjyw commented on src/python/pants/binaries/binary_tool.py:
```diff
@@ -70,10 +84,35 @@ def select(self, context=None):
+
+  # can override this and call super() to compose
+  @classmethod
+  def support_dir_paths(cls):
```
Returning path segments seems overly complicated and is confusing.

How about we just eliminate the whole support dir concept entirely, and just do a one-time rearrangement of our binary tools so that the role of support dir is fulfilled by the binary name.

E.g., why do we need to have `protoc` under `bin/protobuf` when we can have it under `bin/protoc` ? And we can get rid of the `script` subdir entirely, and put everything under `bin`. There's no need to distinguish between them via path prefix.

10 minutes of S3 copying could simplify this code a lot.

cosmicexplorer replied:
I agree that it is confusing, I spent a lot of time thinking about the right thing to do here and I wasn't super satisfied with this, but I didn't know we could do what you described. I don't know how easy modifying S3 to make this work would be, but I agree otherwise that sticking everything in the `bin/` directory is a great idea and would remove many lines of code here.

benjyw replied:
Should probably run this by @stuhood and @jsirois, but I think it should be fine to just change the structure in our S3 dirs. We'd keep the old structure to support existing Pants versions, of course. I have relevant AWS permissions, so I can take a crack at it.

cosmicexplorer replied:
Additionally, requiring the `options_scope` to be the same as the directory the binary is stored under is what I really wanted to do, but I also thought that was off limits, so if we can rearrange the structure of pants binaries easily enough that would also reduce some unnecessary complexity.

benjyw replied:
That's a little less robust, as the option scope can change via a deprecation cycle, and the deprecator may not think to modify S3. Whereas the binary name is fixed.

In some cases we can't have the options_scope be the same as the name for legacy reasons (e.g., the scope of `buildozer` is `buildozer-binary` because `buildozer` is the scope of the task that runs the buildozer binary).

cosmicexplorer replied:
That makes perfect sense -- I didn't see that comment in time! I'll add back the ability to customize the tool's containing directory separately from `options_scope`.

benjyw replied:
I think the directory should just be `bin/<tool_name>`, e.g., `bin/protoc`.

benjyw replied:
In most cases they already are, I bet...

benjyw replied:
OK, it turns out things are a little more complicated - in some cases the "binary name" for the purpose of fetching is, e.g., `go.tar.gz`.  But I think your change will already do the right thing here, ya?

benjyw replied:
And we can get rid of the repetition of archive unpacking code in go_distribution.py and elsewhere.

benjyw replied:
OK, I did the relevant S3 copying - really just moving the `scripts/` to `bin/`, and copying protobuf->protoc, everything else under `bin/` was already in the appropriately named dir.

So now you can hardwire the supportdir BinaryTool passes to BinaryUtils to `bin/<toolname>` and it should work!

benjyw commented:
BTW I just merged https://github.com/pantsbuild/pants/pull/5462 which may give you merge conflicts.

benjyw commented on src/python/pants/binaries/binary_tool.py:
```diff
@@ -5,23 +5,26 @@
   """
   # Subclasses must set these to appropriate values for the tool they define.
   # They must also set options_scope appropriately.
   support_dir = None
```
I think the whole point was to get rid of this field, no? Since it is now always `'bin/{}'.format(self._get_name())`

cosmicexplorer replied:
Oh, I misinterpreted your change to BuildozerBinary -- I think the right thing for now is probably to do exactly that, but allow subclasses to specify a complete `support_dir` as with (current) buildozer if necessary.

benjyw replied:
How about a get_support_dir() classmethod that returns  'bin/{}'.format(cls._get_name()) by default but can be overridden in specific subclasses?

cosmicexplorer replied:
Somehow forgot to refresh and missed your comment again -- I did this in `_resolve_binaryutil_supportdir()` in c60d531.

cosmicexplorer commented:
I merged in the changes from #5462 and refactored the maze of `select_*()` functions in a way that I think is much easier to follow. I also added more relevant (I thought) docstrings to those functions -- let me know if this is more or less readable than before.

benjyw commented on src/python/pants/binaries/binary_tool.py:
```diff
@@ -5,23 +5,30 @@
   :API: public
   """
+  # All subsystems inheriting from BinaryToolBase will go under
+  # bin/`support_subdir` unless they specify an explicit `support_dir`.
```
Seriously, please, let's get rid of this unnecessarily confusing concept. Or at least demote it as I suggested, via a classmethod. If someone does want to override it, we may as well let them override the `bin` part too - for all we know that may be why they need to.

benjyw replied:
OK I see that this is possible via setting `support_dir`, but this is just confusing. Let's kill both `support_dir` and `support_subdir` entirely. We can achieve the same amount of flexibility with a classmethod that uses the value of `get_name()` by default, and can be overridden for custom logic. This is too declarative...

cosmicexplorer replied:
Ok, I wasn't paying enough attention before, and I didn't realize your comments applied to this as well. Sorry about that. That makes sense to me now, and I've implemented it and changed the method's name to `get_support_dir()`.

cosmicexplorer replied:
I really need to refresh pages -- I can implement the suggestion to use `get_name()` and allow overrides right now.

cosmicexplorer replied:
Sorry for making so many commits in a row -- 83ffdda should be what we were trying to do.

benjyw commented on contrib/buildrefactor/src/python/pants/contrib/buildrefactor/buildozer_binary.py:
```diff
@@ -19,13 +19,16 @@
   replaces_scope = 'buildozer'
   replaces_name = 'version'
 
+  # TODO: Move this to bin/buildozer - buildozer is a native binary.
```
I copied everything from scripts/buildozer to bin/buildozer, so this should no longer be needed. May be worth removing in a separate change though.

cosmicexplorer replied:
I'll make a separate change after this gets merged removing the overrides and TODOs!

benjyw commented on src/python/pants/backend/codegen/protobuf/subsystem/protoc.py:
```diff
@@ -10,8 +10,12 @@
   replaces_scope = 'gen.protoc'
   replaces_name = 'version'
+
+  # 'protoc' is also the binary's filename.
```
Ditto, this should no longer be necessary, but might be worth deleting in a separate change.

cosmicexplorer replied:
As above, I'll make a separate change after this gets merged doing that.

benjyw commented on src/python/pants/backend/native/subsystems/llvm.py:
```diff
@@ -0,0 +1,14 @@
+from pants.binaries.binary_tool import NativeTool
+
+
+class LLVM(NativeTool):
```
Boom! I love how succinct this can be now, even with the archive handling.

cosmicexplorer replied:
Yeah, as someone who's not a pants guru yet I thought using BinaryUtil could be a lot simpler (not that it really _needed_ to be), and wasn't sure at all how to do it until @stuhood pointed me to your PR this week.

I was trying to turn the `archive_type` into a mixin or something, but I've hardly ever used python before contributing to pants and it was a bit difficult to think of how to do that. The fact that `BinaryTool#select()` can return either a path to an executable file or an unpacked directory of anything makes me a little queasy, but I think that can be iterated on if/when we start providing much more than just binaries with BinaryTool/Util. I'm thinking of providing headers/libraries in individual BinaryTool subclasses as a shim for the first iteration of C/C++ support, but that won't be difficult enough to require special treatment, I think, especially since that will hopefully be done in a more structured way when we Finally Solve the Problem of Native Dependencies.

benjyw replied:
I think it's fine - it returns a path to an unpacked archive, but where the "archive" is just a single file, it returns that. I think that's perfectly intuitive.

benjyw commented on src/python/pants/binaries/binary_tool.py:
```diff
@@ -78,13 +77,25 @@ def select(self, context=None):
 
   @classmethod
-  def _get_name(cls):
+  def get_name(cls):
```
Why?

cosmicexplorer replied:
No reason at all -- I intuited from a comment you made earlier that said `get_name()` instead of `_get_name()` that that's what you were thinking, and I had the vague feeling that methods that are allowed to be overridden by subclasses are generally named without a leading underscore, based on ~0 python experience. If it's more idiomatic to have the underscore I'll totally put it back.

benjyw replied:
The underscore is a convention for "this is private". Nothing in the language prevents you from calling it from outside, it's just a convention, but it's a useful one nonetheless. And I don't see any need for this to be overridden.  So I think it should be "private", for now.  It's much easier to make something public later, than vice versa.

cosmicexplorer replied:
Ah, in haste I was confusing `_get_name()` with `get_support_dir()`. I've made the change in 0dacd36.

UnrememberMe commented on src/python/pants/binaries/binary_util.py:
```diff
@@ -129,48 +130,67 @@ def __init__(self, baseurls, timeout_secs, bootstrapdir, path_by_id=None):
-  def _select_binary(self, supportdir, version, name):
-    """Selects a binary matching the current os and architecture.
+  def select(self, supportdir, version, name, platform_dependent, archive_type):
+    """Fetches a file, unpacking it if necessary."""
```
he usage based on the codes currently will be slightly different for archive 
  * name of the script without extension
  * will only support one extension per archive type, say only .tgz, no .tar.gz
and non-archive
  * full name of the script / file

cosmicexplorer replied:
Oops, didn't see this comment before. I mentioned above that it kinda bothers me how `select()` currently returns a path to either a binary or an unpacked archive, if the BinaryTool subclass specifies `archive_type`. It already handles constructing the appropriate filename to fetch (e.g. appending `.tar.gz` when `archive_type='tgz'`) and unpacks the archive before returning the path to the unpacked archive. Right now I think handling the inconsistency in returning a path or a directory will require a small amount of extra code, and is something that should be handled in a subsequent PR which uses that feature to evaluate how simple / appropriate the solution would be (which I'll put up after this gets merged).

cosmicexplorer replied:
A task using the `LLVM` subsystem right now would have to know that `select()` is going to return a directory, but that would be clear enough, I think, because we have to specify `archive_type` in the `LLVM` class itself anyway.

benjyw replied:
I think it's OK as-is. We're assuming a naming convention for the paths we fetch from, but that's fine.
